### ISSUE TITLE
Secondary index shouldn't be used before fully built

### ISSUE THREAD

**Issue by nyh** [scylladb/scylla#7963]
When a user uses the `CREATE INDEX` statement to create a secondary index on a table, as soon as this statement returns the index is considered existing, and it will be used in queries that can make use of this index.

The problem is that if the table already has existing data, the index is built in the background and this can take a long time. During that time queries that use the index might work but return incorrect responses - missing a lot or even all of the data.

I think (and this was first raised by @vladzcloudius) that this behavior doesn't make too much sense. As long as the index isn't fully built, we probably shouldn't use it in queries. Queries that need this index for performance should continue to insist on ALLOW FILTERING until this index is fully built, and do the slow filtering as it does without the index.

Another thing we should do is to provide - and document - how users can know when the index the created finishes building.

**Comment by vladzcloudius** [comment]
@nyh Don't you think that the same reasoning applies to MVs as well: the MV should be be available for querying till fully built?

**Comment by nyh** [comment]
@vladzcloudius I don't know. There is a subtle difference between the way that an SI and MV are used - an MV is a new table, which the user can decide not to use yet (just like it was not used before the MV was added), while a SI modifies the existing table and can break queries that worked before. I am still not sure if a partially-built index will only break "ALLOW FILTERING" requests (which are supposed to be avoided anyway) or is it possible to break other queries. I'm trying to write a test for this now.

**Comment by vladzcloudius** [comment]
> @vladzcloudius I don't know. There is a subtle difference between the way that an SI and MV are used - an MV is a new table, which the user can decide not to use yet (just like it was not used before the MV was added), while a SI modifies the existing table and can break queries that worked before. I am still not sure if a partially-built index will only break "ALLOW FILTERING" requests (which are supposed to be avoided anyway) or is it possible to break other queries. I'm trying to write a test for this now.

True. However the user needs a way to know when the MV contents is ready to be used, @nyh .
According to what I understand MVs have the same issue as SI - they are "readable" right after the corresponding `CREATE` statement is completed.

And just like with SIs the user needs a mean to know that he/she can start reading from the MV.

**Comment by nyh** [comment]
I added to the top comment an example where an ALLOW FILTERING query can break by adding an index before its fully built, but explained (sort of...) why a query that didn't use ALLOW FILTERING cannot break after adding an index.

**Comment by vladzcloudius** [comment]
> > @vladzcloudius I don't know. There is a subtle difference between the way that an SI and MV are used - an MV is a new table, which the user can decide not to use yet (just like it was not used before the MV was added), while a SI modifies the existing table and can break queries that worked before. I am still not sure if a partially-built index will only break "ALLOW FILTERING" requests (which are supposed to be avoided anyway) or is it possible to break other queries. I'm trying to write a test for this now.
> 
> True. However the user needs a way to know when the MV contents is ready to be used, @nyh .
> According to what I understand MVs have the same issue as SI - they are "readable" right after the corresponding `CREATE` statement is completed.
> 
> Hence the (same?) similar API that would inform that the SI is ready would be very helpful for MVs too.

For the record: there is a way to know if MV is built: one can query `system_distributed.view_build_status` for that.
Thanks, @psarna.

**Comment by nyh** [comment]
In pull request https://github.com/scylladb/scylladb/pull/18993 I added a reproducing test for this issue, `test_secondary_index.py`::`test_unbuilt_index_not_used`. It is an improved version of the test suggested above.
